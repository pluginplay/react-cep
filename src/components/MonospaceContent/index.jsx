import React from 'react'
import { FlexItem } from 'styled-flex-component'
import styled from 'styled-components'

class MonospaceContent extends React.Component {

  onClick (e) {
    e.preventDefault()
    const range = document.createRange()
    const sel = window.getSelection()
    range.selectNodeContents(e.target)
    sel.removeAllRanges()
    sel.addRange(range)
  }

  render () {
    return (
      <FlexItem {...this.props} onClick={this.onClick}>
        {this.props.children}
      </FlexItem>
    )
  }

}

export default styled(MonospaceContent)`
  background: ${props => props.theme.background.darker};
  border-radius: 2px;
  padding: 5px;
  margin: 0;
  width: 100%;
  box-sizing: border-box;
  font-family: monospace;
  overflow: auto;
  text-align: left;
`
