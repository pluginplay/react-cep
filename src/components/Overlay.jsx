import Flex from 'styled-flex-component'

export default Flex.extend`
  transition: opacity 0.25s linear, z-index 0s linear ${props => props.visible ? '' : '0.25s'};
  text-align: center;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: ${props => props.theme.background.dark};
  z-index: ${props => props.visible ? 3 : -1};
  opacity: ${props => props.visible ? 1 : 0};
  a {
    color: white;
    text-decoration: none;
    opacity: 0.75;
    transition: opacity 0.2s linear;
    &:hover {
      opacity: 1;
    }
  }
`
