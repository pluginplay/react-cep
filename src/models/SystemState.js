import { observable, action } from 'mobx'

export default class SystemState {
  @observable progress = 0
  @observable progressLabel = 'Loading'
  @observable showProgress = false
  @observable showProgressComplete = false
  @observable errorMessage = null
  @observable errorObject = null
  @observable showErrorOverlay = false
  @observable debugMessage = null
  @observable showDebugOverlay = null
  @observable confirmationMessage = null
  @observable confirmationCallback = null
  @observable showConfirmationOverlay = false

  @action setProgress (progress, label, showProgress = true) {
    this.showProgressComplete = false
    this.showProgress = showProgress
    this.progressLabel = label
    this.progress = progress
  }

  @action progressComplete () {
    this.progress = 100
    this.showProgressComplete = true
    setTimeout(() => {
      this.hideProgress()
    }, 3000)
  }

  @action hideProgress () {
    this.showProgress = false
  }

  @action setErrorMessage (errorText, error) {
    this.errorMessage = errorText
    this.errorObject = error
    this.showErrorOverlay = true
  }

  @action hideErrorMessage () {
    this.errorMessage = ''
    this.errorObject = null
    this.showErrorOverlay = false
  }

  @action addDebugMessage (message) {
    console.log(`[DEBUG] ${message}`)
    if (this.debugMessage === null) {
      this.debugMessage = message
    } else {
      this.debugMessage += `\n${message}`
    }
  }

  @action showDebug () {
    this.showDebugOverlay = true
  }

  @action clearDebug () {
    this.showDebugOverlay = false
    this.debugMessage = null
  }

  @action showConfirmation (message, callback) {
    this.showConfirmationOverlay = true
    this.confirmationMessage = message
    this.confirmationCallback = callback
  }

  @action hideConfirmation () {
    this.showConfirmationOverlay = false
    this.confirmationMessage = null
    this.confirmationCallback = null
  }
}
