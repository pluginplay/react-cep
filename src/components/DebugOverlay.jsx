import React from 'react'
import { observer } from 'mobx-react'
import PropTypes from 'prop-types'
import SystemState from '../models/SystemState'
import Overlay from './Overlay'
import ColorHeader from './ColorHeader'
import OverlayContent from './OverlayContent'
import MonospaceContent from './MonospaceContent'

@observer
class DebugOverlay extends React.Component {
  constructor (props) {
    super(props)
    this._onCloseClickedBound = this.onCloseClicked.bind(this)
  }

  onCloseClicked () {
    this.props.systemState.clearDebug()
  }

  render () {
    return (
      <Overlay visible={this.props.systemState.showDebugOverlay} column full>
        <ColorHeader color={'blue'}>
          <span>Debug Information</span>
          <a href={'#'} onClick={this._onCloseClickedBound}>Close</a>
        </ColorHeader>
        <OverlayContent full column>
          <p>Here is the debug information you asked for. Please send this to the plugin maintainers if
            they ask for it.</p>
          <MonospaceContent grow={1}>{this.props.systemState.debugMessage}</MonospaceContent>
        </OverlayContent>
      </Overlay>
    )
  }
}

DebugOverlay.propTypes = {
  systemState: PropTypes.instanceOf(SystemState).isRequired
}

export default DebugOverlay
