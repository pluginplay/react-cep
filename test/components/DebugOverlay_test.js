import React from 'react'
import { shallow } from 'enzyme'
import { expect, withTheme } from '../setup'
import renderer from 'react-test-renderer'
import SystemState from '../../src/models/SystemState'
import DebugOverlay from '../../src/components/DebugOverlay'

import Overlay from '../../src/components/Overlay'
import MonospaceContent from '../../src/components/MonospaceContent'

let systemState = null
let clearDebugSpy = null

beforeEach(() => {
  clearDebugSpy = jest.spyOn(SystemState.prototype, 'clearDebug')
  systemState = new SystemState()
  systemState.showDebugOverlay = true
  systemState.debugMessage = 'foobar'
})

afterEach(() => {
  clearDebugSpy.mockRestore()
})

const getWrapper = () => {
  return shallow(<DebugOverlay systemState={systemState} />)
}

it('is only visible when showDebugOverlay is true', () => {
  const wrapper = getWrapper()
  expect(wrapper.find(Overlay).first()).to.have.prop('visible').equal(true)
})

it('hides the debug overlay when close is clicked', () => {
  const wrapper = getWrapper()
  wrapper.find('a').findWhere(a => a.text() === 'Close').first().simulate('click')
  jestExpect(clearDebugSpy).toHaveBeenCalledTimes(1)
})

it('displays the proper debug text', () => {
  const wrapper = getWrapper()
  expect(wrapper.find(MonospaceContent).first().children().first()).to.have.text('foobar')
})

it('renders the proper HTML structure', () => {
  const tree = renderer
    .create(withTheme(<DebugOverlay systemState={systemState} />))
    .toJSON()
  jestExpect(tree).toMatchSnapshot()
})
