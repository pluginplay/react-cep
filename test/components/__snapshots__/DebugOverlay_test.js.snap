// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the proper HTML structure 1`] = `
<div
  className="flex-bfOkaE bINiYi"
>
  <div
    className="flex-hMftNG bQfNHT"
    color="blue"
  >
    <span>
      Debug Information
    </span>
    <a
      href="#"
      onClick={[Function]}
    >
      Close
    </a>
  </div>
  <div
    className="flex-hWOzxa dsHpRg"
  >
    <p>
      Here is the debug information you asked for. Please send this to the plugin maintainers if they ask for it.
    </p>
    <div
      className="sc-bdVaJa hcHUEK item-bUAMdM bMtFws"
      onClick={[Function]}
    >
      foobar
    </div>
  </div>
</div>
`;
